subprojects/core/src/main/java/org/gradle/execution/plan/PersistentList.java

```diff
@@ -0,0 +1,71 @@
+/*
+ * Copyright 2021 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.execution.plan;
+
+import org.gradle.internal.Cast;
+
+import javax.annotation.CheckReturnValue;
+import java.util.function.Consumer;
+
+public interface PersistentList<T> {
+    static <T> PersistentList<T> of() {
+        return Cast.uncheckedCast(Nil.NIL);
+    }
+
+    static <T> PersistentList<T> of(T first) {
+        return PersistentList.<T>of().add(first);
+    }
+
+    void forEach(Consumer<? super T> consumer);
+
+    @CheckReturnValue
+    PersistentList<T> add(T element);
+}
+
+class Nil<T> implements PersistentList<T> {
+    public static final Nil<Object> NIL = new Nil<>();
```

Comment:
This should be `Nil<?>`, no?

Reply:
Why should it be `Nil<?>`?

Reply:
Actually, why does `Nil` have a generic type at all? I think what we should have is `Nil.of()` returning a `PersistentList<T>`. That's what people should be referring to instead of `Nil.NIL` (which should be called `INSTANCE` and be `private`).

Reply:
We have `PersistentList.of()` which returns `NIL`. People shouldn't use `Nil` at all.